File: 단계별로 풀어보기/문자열/1157/단어 공부.cpp
#include <iostream>
using namespace std;
//아스키코드 a는 97, A는 65

int* countAlphabet(int* alphabet, char* word){
    for(int i=0; word[i]!='\0'; i++){
        if((int)word[i] >= 97){
            alphabet[(int)word[i]-97]++;
        }
        else{
            alphabet[(int)word[i]-65]++;
        }
    }
    
    return alphabet;
}

int findMaxNum(int* alphabet){
    int max=0;
    bool overlap=false;
    
    for(int i=1; i<26; i++){
        if(alphabet[max]<alphabet[i]){
            max=i;
            overlap=false;
        } else if(alphabet[max]==alphabet[i]){
            overlap=true;
        }else{
            continue;
        }
    }
    
    if(overlap==true)
        return -1;
        
    return max+65;
}

int main(){
    char word[1000001];
    int alphabet[26];
    cin>>word;
    fill_n(alphabet,26,0);
    
    countAlphabet(alphabet, word);
    
    if(findMaxNum(alphabet)==-1){
        cout<<'?';
    }else{
        cout << (char)findMaxNum(alphabet);
    }
}

File: 단계별로 풀어보기/문자열/1157/단어 공부.py
#아스키코드 a는 97, A는 65

def initializeList():
    alphabet = []
    for i in range(26):
        alphabet.append(0)
        
    return alphabet

def countAlphabet(word):
    alphabet = initializeList()
    
    for i in word:
        if(ord(i)>=97):
            alphabet[ord(i)-97]+=1
        else:
            alphabet[ord(i)-65]+=1
    
    return alphabet;
    
def findMaxNum(alphabet):
    max=0
    overlap = False
    
    for i in range(1,26):
        if(alphabet[i]>alphabet[max]):
            max=i
            overlap=False
        elif(alphabet[i]==alphabet[max]):
            overlap=True
        else:
            continue
        
    if overlap == True:
        return -1
    else:
        return max

def printResult(alphabet):
    max = findMaxNum(alphabet)
    if max == -1:
        print('?')
    else:
        print(chr(max+65))
        
    return 0

def init():
    word=list(input().rstrip())
    alphabet=countAlphabet(word)
    printResult(alphabet)

init()
    
